### ISSUE TITLE
ZodObject not inferred when using ZodSchema

### ISSUE THREAD

**Issue by prescience-data** [colinhacks/zod#858]
I'm trying to extend a `ZodObject` schema which is bound to an existing interface with `ZodSchema`, however the schema is detected as `ZodType` which does not have the `extend()` method.

Is there some way to specify that the `ZodSchema` is specifically a `ZodObject`?

```typescript

ActiveServiceModel: ZodSchema<ActiveService> = z.object({
  service_id: z.string(),
  team_id: z.string().uuid()
})

const ExtendedActiveServiceModel = ActiveServiceModel.extend({
  bar: z.string()
})

```

### Error Detail

```
TS2339: Property 'extend' does not exist on type 'ZodType '.
(ActiveServiceModel: z.ZodType<ActiveService, z.ZodTypeDef, ActiveService>)
```

**Comment by fnky** [comment]
You can use `ZodObject<AuthService>` in place of ZodSchema. This would also correctly reflect the true type of `z.object`.

**Comment by prescience-data** [comment]
This doesn't appear to be the case:

```typescript

export const ActiveServiceModel: ZodObject<ActiveService> = z
  .object({
    service_id: z.string().max(100),
    team_id: z.string().uuid()
  })

```

reports an error:

```
TS2344: Type 'ActiveService' does not satisfy the constraint 'ZodRawShape'.   Index signature for type 'string' is missing in type 'ActiveService'.
```

The signature for `ZodObject` appears to require the entire schema to be recapitulated in the type definition (as the `Output` signature placed _after_ the `T` definition.

```typescript
 ZodObject<T,UnknownKeys,Catchall,Output,Input> extends ZodType
```

If I substitute a generic `ZodRawShape` into `T` it appears to break a lot of inferences, and the only way to get past `T` is to fully redefine the Zod schema shape.

**Comment by scotttrinh** [comment]
Yeah, there is not an easy way to extend a `ZodSchema`, and as you've found constructing a `ZodObject` is non-trivial and full of duplication.

You will find that requiring a schema to conform to an interface will add friction like this in a lot of places, so I would caution against doing it unless there is a really good reason that is worth the trade offs. Zod has been designed to be the source of truth for the types, and while there are definitely valid cases for wanting to use something like `ZodSchema` (validating third party types for instance), it is fraught with issues like this.

If it's possible to let your schema's be the source of truth, that will resolve your issue. If not, I would suggest perhaps trying to use a shared "shape" rather than using `extend` like this:

[CodeSandbox](https://codesandbox.io/s/stoic-butterfly-jbqn2?file=/src/index.ts)
```ts
interface ActiveService {
  service_id: string;
  team_id: string;
}

const shape = {
  service_id: z.string(),
  team_id: z.string().uuid()
};

const ActiveServiceModel: z.ZodSchema<ActiveService> = z.object(shape);

const ExtendedActiveServiceModel = z.object({
  ...shape,
  bar: z.string()
});
```

—At the very least, you'll start getting errors if the shape and interface get out of sync. But extending this beyond the module boundary is pretty messy and unergonomic.

**Comment by prescience-data** [comment]
For reference the use case is ensuring the Zod schemas match the generated types from a Postgres database. 

Ultimately all that needs to occur is that build or CI should fail if the types are out of sync. 

@scotttrinh  I think I can possibly see a solution based on your example above (given that a build-time failure is all I technically require), thanks for that - I'll check and report results.

**Comment by JRGranell** [comment]
Hi, I have a similar issue and needed to type `ZodObject` from external type. I found this worked, although it seems straightforward it probably means I've missed something!

```
import { z } from 'zod'

type Properties<Input> = Required<{
        [K in keyof Input]: z.ZodType<Input[K], any, Input[K]>;
}>

type Range = {
    min: number;
    max: number;
}

const schema = z.object<Properties<Range>>({
   min: z.number(),
   max: z.number()
});
```

I've found that TS (I'm using v4.5.4) will error if a property is missing or mis-spelt or has an incorrect property type.

**Comment by prescience-data** [comment]
@JRGranell 
this is very cool:
```typescript

type Properties<Input> = Required<{
        [K in keyof Input]: z.ZodType<Input[K], any, Input[K]>;
}>
```

thanks for that 🤙🏻

**Comment by JacobWeisenburger** [comment]
This seems like this question was answered. May I close this?
